sbi/utils/posterior_ensemble.py

```diff
@@ -0,0 +1,446 @@
+            # for SNPE
+            priors = [posterior.prior for posterior in self.posteriors]
+        except AttributeError:
+            # for SNLE and SNRE
+            priors = [posterior.proposal for posterior in self.posteriors]
+
+        # assert same type
+        same_type = all(isinstance(prior, type(priors[0])) for prior in priors)
+
+        def compare_params(dist_x, dist_y):
+            all_equal = True
+            for x, y in zip(dist_x.__dict__.values(), dist_y.__dict__.values()):
+                if type(x) == Tensor:
+                    all_equal = all_equal and torch.equal(x, y)
+                else:
+                    all_equal = all_equal and (x == y)
+            return all_equal
+
+        if same_type:
+            # assert same parameters
+            same_params = all(compare_params(prior, priors[0]) for prior in priors)
+
+        same_prior = same_type and same_params
+        if same_prior:
+            return priors[0]
+        else:
+            return None
+
+    def ensure_same_device(self, posteriors):
+        """Ensures that all posteriors in the ensemble are on the same device.
+
+        Args:
+            posteriors: List containing the trained posterior instances that will make
+                up the ensemble.
+
+        Raises:
+            AssertionError if ensemble components have different device variables.
+
+        Returns:
+            A device string, that is the same for all posteriors.
+        """
+        devices = [posterior._device for posterior in posteriors]
+        assert all(
+            device == devices[0] for device in devices
+        ), "Only supported if all posteriors are on the same device."
+        return devices[0]
+
+    @property
+    def weights(self):
+        return self._weights
+
+    @weights.setter
+    def weights(self, weights: List[float] or Tensor):
+        """Set relative weight for each posterior in the ensemble.
+
+        Weights are normalised.
+
+        Args:
+            weights: Assignes weight to each posterior distribution.
+        """
+        if weights is None:
+            self._weights = torch.tensor(
+                [1.0 / self.num_components for _ in range(self.num_components)]
+            )
+        elif type(weights) == Tensor or type(weights) == List:
+            self._weights = torch.tensor(weights) / sum(weights)
+        else:
+            raise TypeError
+
+    def sample(self, sample_shape: Shape = torch.Size(), **kwargs) -> Tensor:
+        r"""Return samples from posterior ensemble.
+
+        All kwargs are passed directly through to `posterior.sample()`.
+
+        Args:
+            sample_shape: Desired shape of samples that are drawn from posterior
+                ensemble. If sample_shape is multidimensional we simply draw
+                `sample_shape.numel()` samples and then reshape into the desired shape.
+
+        Returns:
+            Samples drawn from the ensemble distribution.
+        """
```

Comment:
Here, you have to call the [_x_else_default_x](https://github.com/mackelab/sbi/blob/8657c55b39ec8f673a0eaa0cb2428b7f2ce5b14f/sbi/inference/posteriors/base_posterior.py#L134-L145) method. If `x=None`, default_x will be used.

Reply:
I thought I could just rely on the class method of the individual posteriors that gets called downstream anyway?